compiler/rustc_const_eval/src/interpret/intern.rs

```diff
@@ -159,6 +169,13 @@ pub fn intern_const_alloc_recursive<
             && inner_mutability == Mutability::Not
             && !prov.immutable()
         {
+            if base_alloc_id == alloc_id && is_static {
```

Comment:
If you move this check up out of the `if`, I think you don't need to do the `GlobalAlloc::Memory(_)` check below?